generators/importanceSampler.cc

```diff
@@ -0,0 +1,423 @@
+			return; // apply lag
+		}
+
+		bool           valid;
+		double         tPrime;
+		TLorentzVector pBeam;
+		TLorentzVector pX;
+#ifdef _OPENMP
+#pragma omp critical(prodKin)
+#endif
+		{
+			boost::tuples::tie(valid, tPrime, pBeam, pX) = getProductionKinematics(parameters[3*_nPart-6]);
+		}
+		if (not valid) {
+			printErr << "generation of production failed" << std::endl;
+			throw;
+		}
+
+		const TLorentzRotation toLab = rpwa::isobarAmplitude::gjTransform(pBeam, pX).Inverse();
+		std::vector<TVector3> decayKinMomenta(_nPart);
+		for (size_t part = 0; part < _nPart; ++part) {
+			TLorentzVector p = nBodyPhaseSpace.daughter(part);
+			p.Transform(toLab);
+			decayKinMomenta[part] = p.Vect();
+		}
+
+		std::vector<double> var;
+		if (_storeMassAndTPrime) {
+			var.push_back(pX.M());
+			var.push_back(tPrime);
+		}
+		std::vector<TVector3> prodKinMomenta(1, pBeam.Vect());
+#ifdef _OPENMP
+#pragma omp critical(fileWriter)
+#endif
+		{
+			_fileWriter.addEvent(prodKinMomenta, decayKinMomenta, var);
+		}
+	}
+
+	timerTot.Stop();
+#ifdef _OPENMP
+#pragma omp atomic
+#endif
+	_funcCallInfo[CALCULATEOBSERVABLES].totalTime += timerTot.RealTime();
+}
+
+
+bool
+rpwa::importanceSampler::initializeFileWriter(TFile*             outFile,
+                                              const std::string& userString,
+                                              const bool         storeMassAndTPrime,
+                                              const std::string& massVariableName,
+                                              const std::string& tPrimeVariableName)
+{
+	_storeMassAndTPrime = storeMassAndTPrime;
+
+	std::vector<std::string> prodKinParticleNames(1);
+	prodKinParticleNames[0] = _beam.particle.name();
+	std::vector<std::string> decayKinParticleNames(_finalState.particles.size());
+	for (size_t part = 0; part < _nPart; ++part) {
+		decayKinParticleNames[part] = _finalState.particles[part].name();
+	}
+
+	std::map<std::string, std::pair<double, double> > binning;
+	binning[massVariableName]   = _massAndTPrimePicker->massRange();
+	binning[tPrimeVariableName] = _massAndTPrimePicker->tPrimeRange();
+	std::vector<std::string> var;
+	if (_storeMassAndTPrime) {
+		var.push_back(massVariableName);
+		var.push_back(tPrimeVariableName);
+	}
+
+	const bool valid = _fileWriter.initialize(*outFile,
+	                                          userString,
+	                                          rpwa::eventMetadata::REAL,
+	                                          prodKinParticleNames,
+	                                          decayKinParticleNames,
+	                                          binning,
+	                                          var);
+
+	return valid;
+}
+
+
+bool
+rpwa::importanceSampler::finalizeFileWriter()
+{
+	return _fileWriter.finalize();
+}
+
+
+boost::tuples::tuple<bool, double, TLorentzVector, TLorentzVector>
+rpwa::importanceSampler::getProductionKinematics(const double xMass) const
```

Comment:
is there a way to avoid the code duplication from `diffractivePhaseSpace::event()`?